### DISCUSSION TITLE
RomX Video ROM

### DISCUSSION THREAD

**Discussion by computer-logix** [markadev/AppleII-VGA#28] [Q&A] [answered]
I installed a RomX on my Apple IIe and was going crazy trying to figure out why the Video ROM wouldn't use any of the optional fonts until it hit me that I was using the Pico for VGA video.   I switched to the AV cord and the fonts were there all along.

Are there plans to support hardware fonts from ROM in the future?   If not, how do you think it would be accomplished?   I haven't had a chance yet to look at the source yet, but if anyone more experienced with this project had an idea of where to start it would be appreciated!

**Comment by markadev** [comment, marked as answer]
Hey there! Ah yeah, since the Pico is emulating the entire video output circuitry it has its own copy of the video ROM internally to generate the text characters.

When you say "support hardware fonts from ROM" so you mean supporting fonts from the other video ROM banks on the RomX? That would not be possible to implement with this VGA card's design since the video ROM is only accessible to the video generator circuitry on the motherboard.  The same issue would arise if you swapped the default video ROM for one with alternate characters; there's no way for an expansion card to access the video ROM over the Apple bus.

But, the VGA firmware does allow characters in the video ROM to be redefined [from software using registers](https://github.com/markadev/AppleII-VGA/blob/main/docs/Usage.md#registers-2--3---text-character-pattern-offset--data) so you could load a custom font into the Pico that matches one of the fonts in your RomX using a basic or assembly program.  Switching fonts definitely won't be as slick as the RomX's interface because you'd have to run the loader program you write, but it can be done!  I've considered adding more character ROM slots, like the RomX or the AnalogV2, so that they can be loaded and switched easier but I just haven't had the need for it yet.

To add support for multiple character set slots I think it just comes down to designing how to store them persistently in the Pico's flash and then adding some registers to trigger loading & saving into the [active character ROM that the renderer is using](https://github.com/markadev/AppleII-VGA/blob/7c5226396e736f15647550f0bd85da18e72c4622/pico/buffers.c#L18).  The rendering code would just take it from there with no additional changes needed!

**Reply by computer-logix** [reply]
Makes sense.    I took a quick peek at the source and found the bitmaps.    If I can find the time I might work on the changes you suggested.   Thanks for your input!

**Reply by markadev** [reply]
Hey there, not sure if you still intend to look at this (it's fine if not!) but I've just added support for some of the more standard built-in video ROMs: https://github.com/markadev/AppleII-VGA/pull/31

Was there a specific character set you were looking to use, or do you use a custom made character set?  I'm kinda curious how other folks customize their Apple II's in this area.

**Comment by computer-logix** [comment]
No specific font.   I just noticed that if I initialized an alternate 
font on the romx boot rom it doesn't show it using the pico video method 
and it shows fine on the composite video.    Good work!


------ Original Message ------
From ***@***.*** (Mark Aikens)" ***@***.***>
To "markadev/AppleII-VGA" ***@***.***>
Cc "Richard Bourque" ***@***.***>; "Author" 
***@***.***>
Date 1/15/2024 8:09:42 AM
Subject Re: [markadev/AppleII-VGA] RomX Video ROM (Discussion #28)
